src/tesk_core/filer.py

```diff
@@ -417,47 +428,158 @@ def ftp_make_dirs(ftp_connection, path):
-                                        .format('HOST_BASE_PATH', 'CONTAINER_BASE_PATH')
-                                      )
-    
-    
-    if   scheme == 'ftp'                : return FTPTransput
-    elif scheme == 'file'               : return fileTransputIfEnabled()
-    elif scheme in ['http', 'https']    : return HTTPTransput
+            raise FileProtocolDisabled("'file:' protocol disabled\n"
+                                       "To enable it, both '{}' and '{}' environment variables must be defined."
+                                       .format('HOST_BASE_PATH', 'CONTAINER_BASE_PATH')
+                                       )
+
+    if scheme == 'ftp':
+        return FTPTransput
+    elif scheme == 'file':
+        return fileTransputIfEnabled()
+    elif scheme in ['http', 'https']:
+        if 's3' in netloc:
+            return S3Transput
+        return HTTPTransput
+    elif scheme == 's3':
+        return S3Transput
     else:
         raise UnknownProtocol("Unknown protocol: '{scheme}'".format(**locals()))
-    
+
+
+class S3Transput(Transput):
+    def __init__(self, path, url, ftype):
+        Transput.__init__(self, path, url, ftype)
+        self.bucket, self.file_path = self.get_names()
+
+    def __enter__(self):
+        client = boto3.resource('s3', endpoint_url=self.extract_endpoint())
+        if self.check_if_bucket_exists(client):
+            sys.exit(1)
+        self.bucket_obj = client.Bucket(self.bucket)
+        return self
+
+    def check_if_bucket_exists(self, client):
+        try:
+            client.meta.client.head_bucket(Bucket=self.bucket)
+        except botocore.exceptions.ClientError as e:
+            # If a client error is thrown, then check that it was a 404 error.
+            # If it was a 404 error, then the bucket does not exist.
+            logging.error('Got status code: %s', e.response['Error']['Code'])
+            if e.response['Error']['Code'] == "404":
+                logging.error("Failed to fetch Bucket, reason: %s", e.response['Error']['Message'])
+            return 1
+        return 0
+
+    def get_names(self):
+        """
+        if the S3 url is similar to s3://idr-bucket-1/README.txt format
+        """
+        if self.url.startswith("s3"):
+            self.url_path = re.sub(r's3:\/', "", self.url)
+
+        """
+        If the s3 url are of following formats
+        * http://mybucket.s3.amazonaws.com/file
+        * http://mybucket.s3-aws-region.amazonaws.com/file
+        * http://s3.amazonaws.com/mybucket/file
+        * http://s3-aws-region.amazonaws.com/mybucket/file
+        """
+
+        match = re.search('^([^.]+).s3', self.netloc)
+        if match:
+            bucket = match.group(1)
+        else:
+            bucket = self.url_path.split("/")[1]
+        file_path = re.sub(r'^\/' + bucket + '\/', "", self.url_path)
+        logging.info(bucket + ", " + file_path)
+        return bucket, file_path
+
```

Comment:
Can you actually give examples of expected behaviour of this function for all different style of S3 URL supported by your code?

Reply:
Added as comments in the source code.